### ISSUE TITLE
Getting different results with filter and get_by

### ISSUE THREAD

**Issue by hakib** [josiahcarlson/rom#58]
I've got a model called `Trip` with a column pk defined as 
```python 
pk = rom.String(required=True, index=True, unique=True)
```

```python
In [127]: t
Out[127]: <models.Trip at 0x6c93c18>

In [128]: t._new
Out[128]: False

In [129]: t.pk
Out[129]: '2704235'

In [130]: Trip.get_by(pk='2704235')

In [132]: Trip.query.filter(pk='2704235').all()
Out[132]: [<models.Trip at 0x4137908>]
```

I'v set null session `rom.util.use_null_session()` to avoid any caching issues. 

I get this everywhere and i'm wondering whether there is any fundemental difference between `get_by` and `filter` that I am missing. If there is an additional configuration to assure that `filter` and `get_by` will always yield the same results please point it out. 

Thanks.

**Comment by josiahcarlson** [comment]
In this case, get_by() and filter() should be returning the same results. I will investigate and let you know when I know what is going on and when you should expect a release that fixes the issue.

**Comment by hakib** [comment]
Thanks man. I've been trying to create a reproducible example to pin down the exact issue but I havent succedded yet. In the meantime i've turned off caching and i'm using only filter. This is the only way i managed to get stable results.

And yes, You guessed it - I'v sub classed `rom.Model` and added a pk column for my logical keys. I would appreciate it if you can alloborate on how I can set the id instead.

Thanks.

**Comment by josiahcarlson** [comment]
Looking at this further, this specific example works for me (`get_by()` returns the one object, `.filter()` returns one object in a list). Did you maybe add the `unique=True` after initially saving the object? Either way, that doesn't mean that there doesn't exist a problem in general - in fact, using a string column to store what looks to be an integer value will result in additional Redis memory usage over using an `Integer()` column with `index=True` and `unique=True`. But at least for this very specific use-case, what you are doing *should* work.

As for being able to override the primary key, this is what I use for my own caching:
```
class MyModel(Model):
    id = PrimaryKey()

MyModel.id = Integer(required=True)
```

There are several caveats when using this method:
* Unique indexes are not supported, though they are sort-of implied by the fact that there is one place to store data with any individual id, but you have to use `.get(id)` and not `.get_by(id=id)`
* If you want to be able to perform numeric range queries, you will need to add `index=True` to both the `PrimaryKey()` definition and the `Integer()` replacement
* If you do `MyModel(id=5)` in one place and `MyModel(id=5)` later, any subsequent `.save()` may override data, and may partly or fully break other indexes - always use a decent locking mechanism to ensure that you're not breaking your indexes (I can explain more about this later, as this might be really important)

So... long story short; you can do this, but it's going to look different, and you need to be careful when updating/writing data to not break some of your indexes. Something like...
1. Acquire lock for object with primary key X
2. `a = RomModel.get(X) or RomModel(pk=X)`
3. Update/modify the object `a`
4. `a.save()`

`rom` doesn't currently have built-in locking functionality. You can use the `LuaLock()` implementation from the standard Python Redis client - it's generally good enough for this kind of thing.

**Comment by josiahcarlson** [comment]
Actually, what I do looks a bit more like...

```
class MyModel(Model):
    id = PrimaryKey()

MyModel._columns['id'].__class__ = Integer
```

Because neither you nor I are able to reliably reproduce the issue, I'm going to close the bug. If you'd like more help on the caching side of things, or in how to use `rom` for your caching, please feel free to reply here or to email me directly.

**Comment by hakib** [comment]
Hey @josiahcarlson, 
I need to solve this (or at least understand this) because i'm losing faith my results.

I think the problem is in the index. 

``` python
In [144]: Route.get_by(pk='102')

In [145]: Route.query.filter(pk='102').first()
Out[145]: <models.Route at 0x456b8d0>

In [146]: conn = rom.util.get_connection()

In [147]: conn.keys('Route:pk:*')
Out[147]:
['Route:pk:171:idx',
 'Route:pk:102:idx',
... 
 'Route:pk:142:idx',
 'Route:pk:152:idx']

In [148]: conn.hmget('Route:pk:102','idx')
Out[148]: [None]
```

I tried looking for a safe way to drop create the index (tried stuff like setting `Route._columns.get('pk')._index = False` and then `rom.utils.clean_old_index` and `rom.utils.refresh_indices` but it did not drop the index). 


Is there a way to regenerate the index on an existing model to verify that this is an index issue  ? 

Thanks, Haki.

**Comment by josiahcarlson** [comment]
Your string column has both a unique index and a full text index. The problem stems from a few weird interop issues:
1. There is currently no way to tell rom "use the text index", "use the unique index", or "use the numeric index" (if there could be all three at the same time) on different queries
2. Because you are using a string column as a fake primary key column (likely for the unique column specifier - the most recent rom release offers unique indexes in integer columns), `.get_by()...` will use the unique index (why that isn't working for you, I don't know), but `.query.filter(...)` will use the full text index (which seems to work)

Regarding your sample output, of course `conn.hmget('Route:pk:102','idx')` will return `[None]`, there is no `'Route:pk:102'` key. `'Route:pk:102:idx'` is an index of all `Route` instances with a `pk` of `102`; it's an inverted index for plain text searching. And if you didn't understand some part of that, it's okay, it's not important.

Just to understand what you have and what you want (please correct pieces as necessary): you want a fake primary key column to offer a secondary primary key (because you can't really override the `rom` built in one), you want to be able to query by exact value, maybe by range, and you want to be able to ensure that the column value is unique. Aside from the disambiguation in querying, using a column type of: `Integer(index=True, unique=True)` should get you that. Use `.get_by()` for individual entity lookup, and `.query.filter(...)` for range searching.

I will add an API for helping out with "which index" issues during queries. Just let me know if my understanding of what you need is correct.

**Comment by hakib** [comment]
You understand the problem correctly. I will try changing the column to integer with `unique=True, index=True`.

In addition to your suggestion to add more control over "which index" I think some sort of mechanism for rebuilding and maintaining indexes and change in general is also in order (is there already one i am not aware of ?). Something like a poor man's south / django migrations for simple cases like adding an index on an existing model etc..

**Comment by josiahcarlson** [comment]
There already exists a mechanism for rebuilding and maintaining indexes as columns and indexes are added and removed. It is `rom.util.refresh_indices()`, and it is a generator that yields its progress. As an example of its use:

```
for progress, total in rom.util.refresh_indices(MyModel):
    print "%i of %i processed"%(progress, total)
```

I'm in the process of adding some convenience keygens for string/text columns, as well as a warning on string/text columns with indexes without an explicit keygen, to ensure that users get what they are actually interested in with string indexes. My fix for #56 and these additional keygen options will be part of a 0.30.0 release soon.

**Comment by josiahcarlson** [comment]
I just released 0.30.0 with better explicit support for `.get_by()` queries (pass `_numeric=True` to force a numeric range query if both a unique and numeric index already exists on the column), along with a warning for using a String/Text column with `index=True` without an explicit keygen. That, along with Integer columns being able to have `unique=True` should prevent the types of issues you have been facing.

Also, I just caught your mention of `refresh_indices()` not working for you (I don't know how I missed it before). I suspect the issue you were facing with it not dropping/creating indexes was based on not knowing that it was a generator. Hopefully with the usage example provided above, your indexes will be recreated and working again soon :)
